federation/plugin/pkg/admission/schedulingpolicy/admission.go

```diff
@@ -0,0 +1,217 @@
+	if err != nil {
+		return nil, err
+	}
+
+	c := &admissionController{
+		Handler:                  admission.NewHandler(admission.Create, admission.Update),
+		policyEngineClient:       policyEngineClient,
+		policyEngineRetryBackoff: config.RetryBackoff,
+	}
+
+	return c, nil
+}
+
+func (c *admissionController) Validate() error {
+	if c.client == nil {
+		return fmt.Errorf("%s requires a client", pluginName)
+	}
+	return nil
+}
+
+func (c *admissionController) SetInternalKubeClientSet(client internalclientset.Interface) {
+	c.client = client
+}
+
+func (c *admissionController) Admit(a admission.Attributes) (err error) {
+
+	failClosed, err := c.shouldFailClosed()
+	if err != nil {
+		return c.handleError(a, true, err)
+	}
+
+	obj := a.GetObject()
+	decision, err := newPolicyEngineQuery(c.policyEngineClient, c.policyEngineRetryBackoff, a.GetUserInfo(), obj, a.GetKind()).Do()
+
+	if err != nil {
+		return c.handleError(a, failClosed, err)
+	}
+
+	if err := decision.Error(); err != nil {
+		return c.handleError(a, true, err)
+	}
+
+	mergeAnnotations(obj, decision.Resource.Metadata.Annotations)
+
+	return nil
+}
+
+func (c *admissionController) handleError(a admission.Attributes, failClosed bool, err error) error {
+
+	if !failClosed {
+		glog.V(4).Infof("an error occurred but request will be allowed (%v namespace does not contain any ConfigMap resources): %v.", policyConfigMapNamespace, err.Error())
+		return nil
+	}
+
+	c.publishEvent(a, err.Error())
+
+	return admission.NewForbidden(a, err)
+}
+
+func (c *admissionController) publishEvent(a admission.Attributes, msg string) {
+
+	obj := a.GetObject()
+
+	ref, err := ref.GetReference(api.Scheme, obj)
+	if err != nil {
+		glog.Errorf("get reference for %v (kind: %v) failed: %v", a.GetName(), a.GetKind().Kind, err.Error())
+		return
+	}
+
+	event := &api.Event{
+		InvolvedObject: *ref,
+		Message:        msg,
+		Source: api.EventSource{
+			Component: fmt.Sprintf("schedulingpolicy"),
+		},
+		Type: "Warning",
+	}
+
+	if _, err := c.client.Core().Events(a.GetNamespace()).Create(event); err != nil {
+		glog.Errorf("create event for %v (kind: %v) failed: %v", a.GetName(), a.GetKind().Kind, err.Error())
+		return
+	}
+}
+
+// shouldFailClosed returns true if the admission controller should fail-closed
+// in the event of network errors, unexpected policy engine responses, etc. The
+// assumption is if the admin has defined one or more policies (as ConfigMaps
+// in the kube-federation-scheduling-policy namespace) then they expect that
+// policy to be enforced--if the policy cannot be enforced because of some
+// unknown error, the request must be denied.
+func (c *admissionController) shouldFailClosed() (bool, error) {
+
+	lst, err := c.client.Core().ConfigMaps(policyConfigMapNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		return true, err
+	}
+
+	return len(lst.Items) > 0, nil
+}
+
+func loadConfig(file io.Reader) (*admissionConfig, error) {
```

Comment:
Admission controllers seem to implement their own config loading. For example: https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/imagepolicy/admission.go

Most of this code is specific to this admission controller (e.g., admissionConfig type, retry normalization, etc.)